File: src/platforms/facebook.py
"""
Facebook poster via the Graph API.

Supports:
  - Text-only posts
  - Single photo
  - Multi-photo posts (attached as a batch)
  - Video post (link to publicly accessible video URL)

Env vars required:
  FACEBOOK_ACCESS_TOKEN   (Page access token with pages_manage_posts scope)
  FACEBOOK_PAGE_ID

Media can be public URLs or local file paths.
"""

from __future__ import annotations

import os
from pathlib import Path

import requests

from src.utils.spreadsheet import Post

GRAPH = "https://graph.facebook.com/v19.0"


def _page_id() -> str:
    return os.environ["FACEBOOK_PAGE_ID"]


def _token() -> str:
    return os.environ["FACEBOOK_ACCESS_TOKEN"]


def _raise_for(resp: requests.Response) -> None:
    if not resp.ok:
        raise RuntimeError(f"Facebook API error {resp.status_code}: {resp.text}")


def _upload_photo(media: str) -> str:
    """Upload a photo (URL or local path) and return its photo id."""
    p = Path(media)
    if p.exists():
        with open(p, "rb") as f:
            resp = requests.post(
                f"{GRAPH}/{_page_id()}/photos",
                params={"access_token": _token(), "published": "false"},
                files={"source": f},
            )
    else:
        resp = requests.post(
            f"{GRAPH}/{_page_id()}/photos",
            params={"access_token": _token(), "url": media, "published": "false"},
        )
    _raise_for(resp)
    return resp.json()["id"]


def post(p: Post) -> dict:
    """Publish a Post to a Facebook Page. Returns the API response dict."""
    message = p.full_text
    media = p.media

    # ── Video post ────────────────────────────────────────────────────────────
    if len(media) == 1 and media[0].lower().endswith((".mp4", ".mov", ".avi")):
        params = {
            "access_token": _token(),
            "description": message,
            "file_url": media[0],
        }
        resp = requests.post(f"{GRAPH}/{_page_id()}/videos", params=params)
        _raise_for(resp)
        result = resp.json()
        print(f"  [facebook] Video posted: id {result.get('id')}")
        return result

    # ── Single photo ──────────────────────────────────────────────────────────
    if len(media) == 1:
        p_obj = Path(media[0])
        if p_obj.exists():
            with open(p_obj, "rb") as f:
                resp = requests.post(
                    f"{GRAPH}/{_page_id()}/photos",
                    params={"access_token": _token(), "message": message},
                    files={"source": f},
                )
        else:
            resp = requests.post(
                f"{GRAPH}/{_page_id()}/photos",
                params={"access_token": _token(), "url": media[0], "message": message},
            )
        _raise_for(resp)
        result = resp.json()
        print(f"  [facebook] Photo posted: id {result.get('id')}")
        return result

    # ── Multi-photo post ──────────────────────────────────────────────────────
    if len(media) > 1:
        photo_ids = [_upload_photo(m) for m in media[:10]]
        attached = [{"media_fbid": pid} for pid in photo_ids]
        resp = requests.post(
            f"{GRAPH}/{_page_id()}/feed",
            params={"access_token": _token()},
            json={"message": message, "attached_media": attached},
        )
        _raise_for(resp)
        result = resp.json()
        print(f"  [facebook] Multi-photo post: id {result.get('id')}")
        return result

    # ── Text-only post ────────────────────────────────────────────────────────
    resp = requests.post(
        f"{GRAPH}/{_page_id()}/feed",
        params={"access_token": _token(), "message": message},
    )
    _raise_for(resp)
    result = resp.json()
    print(f"  [facebook] Text post: id {result.get('id')}")
    return result

File: web/src/lib/platforms/facebook.ts
import { loadSettings } from "@/lib/settings";

const GRAPH = "https://graph.facebook.com/v19.0";

/** See instagram.ts: Meta's fetcher rejects *.vercel.app URLs with 9004/2207052.
 *  Hand it a drive.google.com CDN URL instead so the host is trusted. */
function toMetaSafeImageUrl(url: string): string {
  const match = url.match(/\/api\/media\/([a-zA-Z0-9_-]+)(?:\.[a-z]+)?(?:\?|$)/i);
  if (!match) return url;
  // lh3 direct (=w1080-rj forces JPEG render). thumbnail endpoint redirects
  // to lh3 and Meta's fetcher won't follow redirects.
  return `https://lh3.googleusercontent.com/d/${match[1]}=w1080-rj`;
}

function toMetaSafeVideoUrl(url: string): string {
  const match = url.match(/\/api\/media\/([a-zA-Z0-9_-]+)(?:\.[a-z]+)?(?:\?|$)/i);
  if (!match) return url;
  return `https://drive.google.com/uc?id=${match[1]}&export=download&confirm=1`;
}

async function resolvePageToken(userOrPageToken: string, pageId: string): Promise<string> {
  // Works whether the stored token is a user token or already a page token
  const resp = await fetch(`${GRAPH}/${pageId}?fields=access_token&access_token=${userOrPageToken}`);
  if (!resp.ok) throw new Error(`Failed to get page token: ${resp.status} ${await resp.text()}`);
  const data = await resp.json();
  return data.access_token || userOrPageToken;
}

async function getPostUrl(postId: string, pageId: string, accessToken: string): Promise<string | undefined> {
  const resp = await fetch(`${GRAPH}/${postId}?fields=permalink_url&access_token=${accessToken}`);
  if (!resp.ok) return undefined;
  const data = await resp.json();
  return data.permalink_url;
}

async function uploadUnpublishedPhoto(
  imageUrl: string,
  pageId: string,
  accessToken: string
): Promise<string> {
  const params = new URLSearchParams({
    access_token: accessToken,
    url: imageUrl,
    published: "false",
  });
  const resp = await fetch(`${GRAPH}/${pageId}/photos?${params}`, { method: "POST" });
  if (!resp.ok) {
    throw new Error(`Facebook photo upload ${resp.status}: ${await resp.text()}`);
  }
  const data = await resp.json();
  if (!data.id) throw new Error("Facebook photo upload returned no id");
  return data.id;
}

export async function postToFacebook(
  text: string,
  imageUrl?: string,
  videoUrl?: string,
  extraImages: string[] = []
): Promise<{ success: boolean; id?: string; url?: string; error?: string }> {
  const settings = loadSettings();
  const { accessToken: storedToken, pageId } = settings.facebook;
  if (!storedToken) return { success: false, error: "Facebook token not configured. Go to Settings to add it." };
  if (!pageId) return { success: false, error: "Facebook page ID not configured. Go to Settings to add it." };

  let accessToken: string;
  try {
    accessToken = await resolvePageToken(storedToken, pageId);
  } catch (e) {
    return { success: false, error: e instanceof Error ? e.message : "Failed to get page token" };
  }

  const allImages = [
    ...(imageUrl ? [imageUrl] : []),
    ...extraImages.filter(Boolean),
  ].map(toMetaSafeImageUrl);

  // Video-only post
  if (videoUrl) {
    const params = new URLSearchParams({
      access_token: accessToken,
      file_url: toMetaSafeVideoUrl(videoUrl),
      description: text,
    });
    const resp = await fetch(`${GRAPH}/${pageId}/videos?${params}`, { method: "POST" });
    if (!resp.ok) {
      const errText = await resp.text();
      return { success: false, error: `Facebook video API error ${resp.status}: ${errText}` };
    }
    const data = await resp.json();
    const url = await getPostUrl(data.id, pageId, accessToken);
    return { success: true, id: data.id, url };
  }

  // Multi-image post: upload each as unpublished, then attach to a single feed post
  if (allImages.length > 1) {
    let photoIds: string[];
    try {
      photoIds = await Promise.all(
        allImages.map((url) => uploadUnpublishedPhoto(url, pageId, accessToken))
      );
    } catch (e) {
      return { success: false, error: e instanceof Error ? e.message : "Facebook multi-image upload failed" };
    }

    const feedParams = new URLSearchParams({
      access_token: accessToken,
      message: text,
      attached_media: JSON.stringify(photoIds.map((id) => ({ media_fbid: id }))),
    });
    const feedResp = await fetch(`${GRAPH}/${pageId}/feed?${feedParams}`, { method: "POST" });
    if (!feedResp.ok) {
      return { success: false, error: `Facebook feed API error ${feedResp.status}: ${await feedResp.text()}` };
    }
    const data = await feedResp.json();
    const url = await getPostUrl(data.id, pageId, accessToken);
    return { success: true, id: data.id, url };
  }

  // Single image post
  if (allImages.length === 1) {
    const params = new URLSearchParams({
      access_token: accessToken,
      url: allImages[0],
      message: text,
      published: "true",
    });
    const resp = await fetch(`${GRAPH}/${pageId}/photos?${params}`, { method: "POST" });
    if (!resp.ok) {
      const errText = await resp.text();
      return { success: false, error: `Facebook API error ${resp.status}: ${errText}` };
    }
    const data = await resp.json();
    // /photos returns { id: photoId, post_id: feedPostId } — use post_id for the timeline post URL
    const feedPostId = data.post_id || data.id;
    const url = await getPostUrl(feedPostId, pageId, accessToken);
    return { success: true, id: feedPostId, url };
  }

  // Text-only post
  const params = new URLSearchParams({
    access_token: accessToken,
    message: text,
  });
  const resp = await fetch(`${GRAPH}/${pageId}/feed?${params}`, { method: "POST" });
  if (!resp.ok) {
    const errText = await resp.text();
    return { success: false, error: `Facebook API error ${resp.status}: ${errText}` };
  }
  const data = await resp.json();
  const url = await getPostUrl(data.id, pageId, accessToken);
  return { success: true, id: data.id, url };
}
